### ISSUE TITLE
Implement fields for multiErr as well as helper for wrap.

### ISSUE THREAD

**Issue by ruseinov** [iov-one/weave#730]
We have several problems with current multierr setup.
1. errors.Append(err, errors.Wrap(err,"some name")) is quite verbose and ugly. 
2. We need to support field tags as otherwise there is no way to see which field caused an error. If we have several fields causing the same type of error - this could lead to tests not covering the code properly.

**Comment by ruseinov** [comment]
some useful discussion here: 
https://github.com/iov-one/weave/pull/727

**Comment by ruseinov** [comment]
In order to Address Wrap:

I think we need to have flexibility here, so we should be able to accept wrapped errors the same we do now and keep the `Append(src, newErr)`, but introduce a method `Appendw(src, new, description, args...)`. This is also consistent with our current approach in the errors package.

**Comment by ruseinov** [comment]
In order to Address tagging:
One of the approaches suggested by @husio was one similar to my initial implementation in that we return a multiErr type from Append to be able to call a method on it, which could be `Tag`.

So the api then is as follows:
`Append(src, dest).Tag("blah")` will add the last added error (if it was not nil) to a bucket of errors that would be returned on "blah". In order for this to work we need to keep track of what was added to the multiErr last and ignore if it was nil. This could be confusing, but it'd work.

And then in tests we'll say something like ByField(multiErr, "field") to get these and test by them. 

Another approach would be to implement AppendTag and AppendwTag to address tagging without any struct exposure and make the api similar to other stuff in errors:
`AppendTag(src, newErr, tag)` `AppendwTag(src, new, tag, description, args...)`. 

We can also go with another approach, which would be to have AppendTag(src, newErr, tag) which does the tagging and `AppendwTag(src, new, tag, args...)` where description is going to be taken from a tag too, so then we have both tag and description having the same value.

**Comment by ruseinov** [comment]
@iov-one/blockchain-team happy to have some discussion here.

**Comment by husio** [comment]
I like `Append(src, dest).Tag("blah")` the most. I think it requires modification and should look more like this:
```go
func Append(errorsOrNil error, errOrNil error, description string) MultiError {
    // ...
}

func (e *MultiError) Tag(name string) error {
}
```
Then the use could look like this
```go
errs = errors.Append(errs, errors.ErrInvalid, "required").Tag("amount")
```

There is a strage case possible, when an error is `nil`. This can be programmed around though.
```go
errs = errors.Append(nil, nil, "yolo").Tag("nothing")
```

**Comment by alpe** [comment]
about 1) I agree that an  `errors.Appendf` would be a nice addition to reduce the boilerplate Wrap calls.
about 2) I think this can be addressed in the test setup by mutating a fixture to match the specific case.

Anyway, to move this discussion further as `Append` was an intermediate step on our way to proper client side errors, I would propose an explicit `Field` error function to address the issue. 
It can provide similar functionality as Wrap with an additional field stored that is later used for grouping.

For example:
```go 
func Field(field string, err error, msg...string)error{
 # msg would be used for wrap functionality but is optional
}
```
This would look like
```go
var err error
if err := msg.Metadata.Validate(); err != nil {
	err = errors.Append(err, errors.Field("metadata", errors.ErrAmount))
}
if fee == nil {
	err = errors.Append(err, errors.Field("fee", errors.ErrAmount, "nil"))
} else {
	err = errors.Append(err, errors.Field("fee", fee.Validate()))
	if !fee.IsNonNegative() {
		err = errors.Append(err, errors.Field("fee",errors.ErrAmount, "negative"))
	}
}
```
This example contains 2 interesting cases that we should cover:
1. nested fields: `metadata.schema`
2. multiple issues on a single field `fee`

As an example for the error result to the client:
```json
{
  "data": {
    "fee": [
      {
        "code": 17,
        "log": "invalid currency"
      },
      {
        "code": 11,
        "log": "amount negative"
      }
    ]
  },
  "metadata.schema": [
    {
      "code": 4,
      "log": "test: invalid message"
    }
  ]
}
```

**Comment by husio** [comment]
I think what @alpe proposed is also good. The example code is good and maybe can be simplified even more by allowing `nil` to be wrapped.
```go
var err error
err = errors.Append(err, errors.Field("metadata", msg.Metadata.Validate()))
// ...
```

I have tried to implement this solution and it is not an easy task. `Field` functionality is easy to understand. The current state of the `errors` package is quite complex internally and requires to keep the whole package in your head when programming. We use type casting in many cases to support edge cases. We have "normal" errors, wrapped errors, stack trace errors, multi errors and now field errors. I think this is too much already and it is getting worse with every next implmentation.
I think as part of this issue we should find a better solution and untangle current relations.

**Comment by ruseinov** [comment]
> We have "normal" errors, wrapped errors, stack trace errors, multi errors and now field errors. I think this is too much already and it is getting worse with every next implmentation.
I think as part of this issue we should find a better solution and untangle current relations.

Either that or we can try and rework multiErr to support fieldError. To me fieldError is just a case of multi err, because otherwise it does not make much sense. On the other hand I agree that implementation could be simpler.

I think we'll have to at least revisit wraps and stack traces to avoid duplicating the code for multierr helpers as we implement this.

**Comment by ethanfrey** [comment]
> We have "normal" errors, wrapped errors, stack trace errors, multi errors and now field errors.

I agree. But let's stop a second. As the package _used_ to be, we are flexible as what we accept as a root cause:

There is stdlib `error`
Then our `Error`, which is only created on registration to hold a abci code.
Or even a `pkg/errors.error` that is passed in

From that we can make any levels of `wrappedError` to hold stack and logs and root cause.

Pretty simple and small API.

-----

Now, we have multiErr and fieldErr and ???
I wonder if we can combine all this into a wrappedError? Or at least the same interface?

As to stack traces. Anything coming from the outside, should first be Wrapped, to ensure we have a wappedError and not stdlib. This will add a stacktrace. Any other error grouping (multi, field, ...) should only accept wrappedError (or some common interface we make marking it as "our" error) and doesn't need to think about stack traces. Just render all sub-errors and they display the various stack traces.

You could add:

```golang
type ourError interface {
    isOurError()
}
```

and set that on our internal error classes, so we know not to process them (add stack) before combining, if that makes it simpler. I have not looked at the code recently, but I don't think there is an inherant reason it should be so tangled

**Comment by husio** [comment]
I think we should avoid having one mega structure that keeps all the information that an error from `weave/errors` can contain. Instead of clubbing everything together I would prefer to have a separate and independent implementation for each functionality. This is also the weakness of the recent implementation - errors package functionalities are no longer independent as they examine for specific types.

I was going in the direction of an error implementation per functionality:
- a static error, that is returned by the `Register` function
- an error wrapper that is attaching a string (`Wrap` and `Wrapf`)
- a stack trace wrapper (attached to `Wrap` and `Wrapf`, possibly other functions as well)
- an error collection (multierror) (`Append` function)
- an error that carries a field/tag information (work in progress)

Because those implementations should not be aware of one another, there must be a common interface that can be utilized. It can be private because it is only for the `errors/` package to use.

Before multi-error implementation was added the [`causer`](https://github.com/iov-one/weave/blob/master/errors/errors.go#L274-L278) interface was the only API all error interfaces shared. Adding multi error complicated this a bit, but I believe we can fix this and continue using an interface instead of sharing the implementation details.

**Comment by ethanfrey** [comment]
This was resolved with #776, correct?

We can close this issue?
